trie/proof.go

```diff
@@ -111,6 +111,189 @@ func (t *StateTrie) Prove(key []byte, proofDb ethdb.KeyValueWriter) error {
+			// trie.resolveAndTrack is not used since in that function the
+			// loaded blob will be tracked, while it's not required here since
+			// all loaded nodes won't be linked to trie at all and track nodes
+			// may lead to out-of-memory issue.
+			blob, err := t.reader.node(prefix, common.BytesToHash(n))
+			if err != nil {
+				log.Error("Unhandled trie error in traverseNodes", "err", err)
+				return nil, err
+			}
+			// The raw-blob format nodes are loaded either from the
+			// clean cache or the database, they are all in their own
+			// copy and safe to use unsafe decoder.
+			tn = mustDecodeNodeUnsafe(n, blob)
+		default:
+			panic(fmt.Sprintf("%T: invalid node: %v", tn, tn))
+		}
+	}
+
+	return tn, nil
+}
+
+func (t *Trie) ProvePath(key []byte, prefixKeyHex []byte, proofDb ethdb.KeyValueWriter) error {
+
+	if t.committed {
+		return ErrCommitted
+	}
+
+	if len(key) == 0 {
+		return fmt.Errorf("key is empty")
+	}
+
+	key = keybytesToHex(key)
+
+	// traverse down using the prefixKeyHex
+	var nodes []node
+	tn := t.root
+	startNode, err := t.traverseNodes(tn, prefixKeyHex, nil) // obtain the node where the prefixKeyHex leads to
+	if err != nil {
+		return err
+	}
+
+	key = key[len(prefixKeyHex):] // obtain the suffix key
+
+	// traverse through the suffix key
+	_, err = t.traverseNodes(startNode, key, &nodes)
+	if err != nil {
+		return err
+	}
+
+	hasher := newHasher(false)
+	defer returnHasherToPool(hasher)
+
+	// construct the proof
+	for _, n := range nodes {
+		var hn node
+		n, hn = hasher.proofHash(n)
+		if hash, ok := hn.(hashNode); ok {
+			enc := nodeToBytes(n)
+			if !ok {
+				hash = hasher.hashData(enc)
+			}
+			proofDb.Put(hash, enc)
+		}
+	}
+
+	return nil
+}
+
+// VerifyPathProof checks the merkle proofs for a given key and root hash. If a prefix key is specified, then it checks
+// the proof against the subtree where the prefix key leads to. It will first reconstruct the trie from the proof and
+// then check if the root hash matches the root hash of the reconstructed trie.
+func VerifyPathProof(rootHash common.Hash, keyHex []byte, prefixKeyHex []byte, proofList [][]byte) (node, error) {
+
+	if len(proofList) == 0 {
+		return nil, fmt.Errorf("proof list is empty")
+	}
+
+	n, err := ConstructTrieFromProof(keyHex, prefixKeyHex, proofList)
+	if err != nil {
+		return nil, err
+	}
+
+	// hash the root node
+	hasher := newHasher(false)
+	defer returnHasherToPool(hasher)
+	hn, _ := hasher.hash(n, true)
+	if hash, ok := hn.(hashNode); ok {
+		if common.BytesToHash(hash) == rootHash {
+			return n, nil
+		}
+	}
+
+	return n, fmt.Errorf("root hash does not match")
+}
+
+// ConstructTrieFromProof constructs a trie from the given proof. It returns the root node of the trie.
+func ConstructTrieFromProof(keyHex []byte, prefixKeyHex []byte, proofList [][]byte) (node, error) {
+	var parentNode node
+	var root node
+
+	keyHex = keyHex[len(prefixKeyHex):]
+
+	for i := 0; i < len(proofList); i++ {
```

Comment:
It may check the child hash proof first, then set the node to the target position.